# Consumer Shopping Behavior Analysis

## 🎯 Business Problem Statement

A leading retail company wants to better understand its customers' shopping behavior in order to improve sales, customer satisfaction, and long-term loyalty.  
The management team has noticed changes in purchasing patterns across demographics, product categories, and sales channels (online vs. offline). 

They want understanding of which factors, such as **discounts**, **reviews**, **seasons**, or **payment preferences** drive consumer decisions and repeat purchases.

---

## ❓ Overarching Business Question

> **"How can the company leverage consumer shopping data to identify trends, improve customer engagement, and optimize marketing and product strategies?"**

---

## 📋 Key Analysis Objectives
1. **Demographic & Pattern Tracking:** Identify buying behaviors across different customer segments.
2. **Channel Performance:** Compare online versus offline purchasing trends.
3. **Feature Impact Analysis:** Quantify how reviews, discounts, and payment methods drive conversion and retention.
4. **Actionable Insights:** Provide data-driven recommendations for marketing and product placement.


## 📝 Dataset & Methodology Notes

- **Data source:** `customer_shopping_behavior_v2.csv`. This project uses a randomized, synthetically generated dataset built to model realistic, non-degenerate relationships. It is **not real transactional data**, and findings should be read as a demonstration of methodology rather than a real business recommendation.
- **Channel data limitation:** the dataset has no explicit online/offline channel field. `shipping_type` is used later as an imperfect proxy (`Store Pickup` vs. everything else) 

## Data Preparation & Modeling (Python)

In [1]:
# Import pandas
import pandas as pd
import numpy as np

# Load dataset from csv
df = pd.read_csv('customer_shopping_behavior_v2.csv')

In [2]:
# Display head to verify data loaded succesfully
df.head()

,Customer ID,Age,Gender,Item Purchased,Category,Purchase Amount (USD),Location,Size,Color,Season,Review Rating,Subscription Status,Shipping Type,Discount Applied,Promo Code Used,Previous Purchases,Payment Method,Frequency of Purchases
0,1,48,Female,Dress,Clothing,33,Vermont,XL,Teal,Fall,3.6,No,Standard,No,No,23,Cash,Bi-Weekly
1,2,39,Male,Belt,Accessories,86,New York,XL,Lavender,Spring,3.9,No,2-Day Shipping,Yes,Yes,16,PayPal,Monthly
2,3,36,Female,Jacket,Outerwear,100,Nebraska,M,Purple,Winter,3.8,Yes,2-Day Shipping,Yes,Yes,4,Bank Transfer,Monthly
3,4,40,Female,Scarf,Accessories,54,Nebraska,L,Yellow,Fall,3.9,No,Next Day Air,No,No,23,Cash,Fortnightly
4,5,61,Male,Jacket,Outerwear,80,Nevada,L,Green,Spring,3.5,No,Store Pickup,No,No,6,Credit Card,Monthly


In [3]:
# Check dataset structure, identify missing data, and verify data types
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3900 entries, 0 to 3899
Data columns (total 18 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Customer ID             3900 non-null   int64  
 1   Age                     3900 non-null   int64  
 2   Gender                  3900 non-null   object 
 3   Item Purchased          3900 non-null   object 
 4   Category                3900 non-null   object 
 5   Purchase Amount (USD)   3900 non-null   int64  
 6   Location                3900 non-null   object 
 7   Size                    3900 non-null   object 
 8   Color                   3900 non-null   object 
 9   Season                  3900 non-null   object 
 10  Review Rating           3863 non-null   float64
 11  Subscription Status     3900 non-null   object 
 12  Shipping Type           3900 non-null   object 
 13  Discount Applied        3900 non-null   object 
 14  Promo Code Used         3900 non-null   

In [4]:
# View summary statistics using .describe()
df.describe(include='all')

,Customer ID,Age,Gender,Item Purchased,Category,Purchase Amount (USD),Location,Size,Color,Season,Review Rating,Subscription Status,Shipping Type,Discount Applied,Promo Code Used,Previous Purchases,Payment Method,Frequency of Purchases
count,3900.000000,3900.000000,3900,3900,3900,3900.000000,3900,3900,3900,3900,3863.000000,3900,3900,3900,3900,3900.000000,3900,3900
unique,NaN,NaN,2,25,4,NaN,50,4,25,4,NaN,2,6,2,2,NaN,6,7
top,NaN,NaN,Male,Sunglasses,Clothing,NaN,South Carolina,M,Yellow,Spring,NaN,No,Standard,No,No,NaN,Debit Card,Quarterly
freq,NaN,NaN,2146,183,1723,NaN,95,1699,180,995,NaN,2879,668,2322,2322,NaN,674,902
mean,1950.500000,43.999487,NaN,NaN,NaN,60.051538,NaN,NaN,NaN,NaN,3.692985,NaN,NaN,NaN,NaN,14.681282,NaN,NaN
std,1125.977353,13.887652,NaN,NaN,NaN,16.261446,NaN,NaN,NaN,NaN,0.598691,NaN,NaN,NaN,NaN,8.579529,NaN,NaN
min,1.000000,18.000000,NaN,NaN,NaN,20.000000,NaN,NaN,NaN,NaN,1.600000,NaN,NaN,NaN,NaN,1.000000,NaN,NaN
25%,975.750000,34.000000,NaN,NaN,NaN,49.000000,NaN,NaN,NaN,NaN,3.300000,NaN,NaN,NaN,NaN,8.000000,NaN,NaN
50%,1950.500000,44.000000,NaN,NaN,NaN,60.000000,NaN,NaN,NaN,NaN,3.700000,NaN,NaN,NaN,NaN,13.000000,NaN,NaN
75%,2925.250000,54.000000,NaN,NaN,NaN,71.000000,NaN,NaN,NaN,NaN,4.100000,NaN,NaN,NaN,NaN,20.000000,NaN,NaN


In [5]:
# Check for missing data/null values
df.isnull().sum()

Customer ID                0
Age                        0
Gender                     0
Item Purchased             0
Category                   0
Purchase Amount (USD)      0
Location                   0
Size                       0
Color                      0
Season                     0
Review Rating             37
Subscription Status        0
Shipping Type              0
Discount Applied           0
Promo Code Used            0
Previous Purchases         0
Payment Method             0
Frequency of Purchases     0
dtype: int64

In [6]:
# Check for exact duplicate rows
print(f"Duplicate rows: {df.duplicated().sum()}")

Duplicate rows: 0


In [7]:
# Replace missing values in Review Rating column with median rating of product category
df['Review Rating'] = df.groupby('Category')['Review Rating'].transform(lambda x: x.fillna(x.median()))

In [8]:
# Check for missing data/null values
df.isnull().sum()

Customer ID               0
Age                       0
Gender                    0
Item Purchased            0
Category                  0
Purchase Amount (USD)     0
Location                  0
Size                      0
Color                     0
Season                    0
Review Rating             0
Subscription Status       0
Shipping Type             0
Discount Applied          0
Promo Code Used           0
Previous Purchases        0
Payment Method            0
Frequency of Purchases    0
dtype: int64

In [9]:
# Modfiy column names to snake_case for readability and consitancy
df.columns = df.columns.str.lower()
df.columns = df.columns.str.replace(' ','_')
df = df.rename(columns={'purchase_amount_(usd)':'purchase_amount'})

In [10]:
# Confirm changes to column names
df.columns

Index(['customer_id', 'age', 'gender', 'item_purchased', 'category',
       'purchase_amount', 'location', 'size', 'color', 'season',
       'review_rating', 'subscription_status', 'shipping_type',
       'discount_applied', 'promo_code_used', 'previous_purchases',
       'payment_method', 'frequency_of_purchases'],
      dtype='object')

In [11]:
# Bin ages into age groups
bins = [17, 24, 34, 49, 71]
labels = ['Young Adult (18-24)', 'Adult (25-34)', 'Middle-aged (35-49)', 'Senior (50+)']
df['age_group'] = pd.cut(df['age'], bins=bins, labels=labels)

In [12]:
# Confirm age_group created and boundaries applied correctly
df[['age','age_group']].head(15)

,age,age_group
0,48,Middle-aged (35-49)
1,39,Middle-aged (35-49)
2,36,Middle-aged (35-49)
3,40,Middle-aged (35-49)
4,61,Senior (50+)
5,33,Adult (25-34)
6,43,Middle-aged (35-49)
7,47,Middle-aged (35-49)
8,67,Senior (50+)
9,39,Middle-aged (35-49)


In [13]:
# Convert frequency_of_purchases into standard day counts stored in a new column
# purchase_frequency_days for readability and analysis consistency
frequency_mapping = {
    'Weekly': 7,
    'Fortnightly': 14,
    'Bi-Weekly': 14,
    'Monthly': 30,
    'Quarterly': 90,
    'Every 3 Months': 90,
    'Annually': 365
}

df['purchase_frequency_days'] = df['frequency_of_purchases'].map(frequency_mapping)

In [14]:
# Confirm purchase_frequency_days created
df[['purchase_frequency_days','frequency_of_purchases']].head(10)

,purchase_frequency_days,frequency_of_purchases
0,14,Bi-Weekly
1,30,Monthly
2,30,Monthly
3,14,Fortnightly
4,30,Monthly
5,365,Annually
6,90,Every 3 Months
7,90,Quarterly
8,30,Monthly
9,90,Quarterly


In [15]:
# discount_applied and promo_code_used appear to be redundant information
df[['discount_applied','promo_code_used']].head(10)

,discount_applied,promo_code_used
0,No,No
1,Yes,Yes
2,Yes,Yes
3,No,No
4,No,No
5,Yes,Yes
6,No,No
7,Yes,Yes
8,No,No
9,No,No


In [16]:
# Compare rows in each column
(df['discount_applied'] == df['promo_code_used']).all()

np.True_

In [17]:
# Drop promo_code_used column as information is redundant to discount_applied
df = df.drop('promo_code_used', axis=1)

In [18]:
# Confirm drop successful
df.columns

Index(['customer_id', 'age', 'gender', 'item_purchased', 'category',
       'purchase_amount', 'location', 'size', 'color', 'season',
       'review_rating', 'subscription_status', 'shipping_type',
       'discount_applied', 'previous_purchases', 'payment_method',
       'frequency_of_purchases', 'age_group', 'purchase_frequency_days'],
      dtype='object')

### Database Connection

In [19]:
%pip install python-dotenv sqlalchemy psycopg2-binary

You should consider upgrading via the 'd:\Program Files\Python\Python310\python.exe -m pip install --upgrade pip' command.


In [20]:
import os
from dotenv import load_dotenv
from sqlalchemy import create_engine
from sqlalchemy import text

# Load environment variables from the .env file
load_dotenv()

# Retrieve the credentials
username = os.getenv("DB_USER")
password = os.getenv("DB_PASSWORD")
host = os.getenv("DB_HOST")
port = os.getenv("DB_PORT")
database = os.getenv("DB_NAME")

# Create the connection engine
engine = create_engine(f"postgresql+psycopg2://{username}:{password}@{host}:{port}/{database}")

with engine.connect() as conn:
    conn.execute(text("DROP VIEW IF EXISTS customer_segments"))
    conn.commit()

# Load DataFrame into PostgreSQL
table_name = "customer"
df.to_sql(table_name, engine, if_exists="replace", index=False)

print(f"Data successfully loaded into table '{table_name}' in database '{database}'.")

Data successfully loaded into table 'customer' in database 'customer_behavior'.


In [21]:
%pip install ipython-sql scipy

Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'd:\Program Files\Python\Python310\python.exe -m pip install --upgrade pip' command.


In [22]:
# Load the SQL extension
%load_ext sql

# Build the connection string from environment variables 
connection_string = f"postgresql+psycopg2://{username}:{password}@{host}:{port}/{database}"
%sql $connection_string
%config SqlMagic.style = '_DEPRECATED_DEFAULT'
%config SqlMagic.displaycon = False
%config SqlMagic.feedback = False

print(f"Connection to '{database}' successful.")

Connection to 'customer_behavior' successful.


## Data Analysis (SQL)

### Reusable customer segmentation view

The New / Returning / Loyal thresholds are used across multiple queries below. Defining as a view  for consistency.

In [23]:
%%sql
CREATE OR REPLACE VIEW customer_segments AS
SELECT *,
       CASE
           WHEN previous_purchases = 1 THEN 'New'
           WHEN previous_purchases BETWEEN 2 AND 10 THEN 'Returning'
           ELSE 'Loyal'
       END AS customer_segment
FROM customer;

[]

## What is the Customer Lifecycle Distribution?

In [24]:
%%sql
SELECT customer_segment,
       COUNT(*) AS number_of_customers,
       ROUND(100.0 * COUNT(*) / SUM(COUNT(*)) OVER(), 1) AS percent_of_total
FROM customer_segments
GROUP BY customer_segment
ORDER BY number_of_customers DESC;

customer_segment,number_of_customers,percent_of_total
Loyal,2430,62.3
Returning,1448,37.1
New,22,0.6


## What is the age group distribution in the Customer Lifecycle?

In [25]:
%%sql
SELECT customer_segment,
       COUNT(CASE WHEN age_group LIKE 'Young Adult%' THEN 1 END) AS young_adult,
       COUNT(CASE WHEN age_group LIKE 'Adult%' THEN 1 END) AS adult,
       COUNT(CASE WHEN age_group LIKE 'Middle-aged%' THEN 1 END) AS middle_aged,
       COUNT(CASE WHEN age_group LIKE 'Senior%' THEN 1 END) AS senior
FROM customer_segments
GROUP BY customer_segment;

customer_segment,young_adult,adult,middle_aged,senior
New,1,2,11,8
Returning,150,258,552,488
Loyal,233,377,941,879


### What is the revenue contribution for each age group?

In [26]:
%%sql
SELECT age_group,
       COUNT(*) AS customers,
       ROUND(AVG(purchase_amount),2) AS avg_spend,
       SUM(purchase_amount) AS total_revenue
FROM customer
GROUP BY age_group
ORDER BY total_revenue DESC;

age_group,customers,avg_spend,total_revenue
Middle-aged (35-49),1504,60.00,90235
Senior (50+),1375,60.13,82683
Adult (25-34),637,59.94,38183
Young Adult (18-24),384,60.16,23100


### What is the revenue contribution by gender?

In [27]:
%%sql
SELECT gender,
       COUNT(*) AS customers,
       ROUND(AVG(purchase_amount),2) AS avg_spend,
       SUM(purchase_amount) AS total_revenue
FROM customer
GROUP BY gender
ORDER BY total_revenue DESC;

gender,customers,avg_spend,total_revenue
Male,2146,60.78,130444
Female,1754,59.15,103757


### Do subscribers or non-subscribers spend more?

In [28]:
%%sql
SELECT subscription_status,
       COUNT(customer_id) AS total_customers,
       ROUND(100.0 * COUNT(customer_id) / SUM(COUNT(customer_id)) OVER(), 1) AS percent_of_customers,
       ROUND(AVG(purchase_amount),2) AS avg_spend,
       ROUND(SUM(purchase_amount),2) AS total_revenue,
       ROUND(100.0 * SUM(purchase_amount) / SUM(SUM(purchase_amount)) OVER(), 1) AS percent_of_revenue
FROM customer
GROUP BY subscription_status
ORDER BY total_revenue desc

subscription_status,total_customers,percent_of_customers,avg_spend,total_revenue,percent_of_revenue
No,2879,73.8,58.28,167785.00,71.6
Yes,1021,26.2,65.05,66416.00,28.4


### Are repeat customers (more than 5 purchases) likely to subscribe?

In [29]:
%%sql
SELECT subscription_status,
       COUNT(customer_id) AS repeat_customers,
       ROUND(100.0 * COUNT(customer_id) / SUM(COUNT(customer_id)) OVER(), 1) AS percent_of_customers
FROM customer
WHERE previous_purchases > 5
GROUP BY subscription_status

subscription_status,repeat_customers,percent_of_customers
No,2443,73.4
Yes,886,26.6


### Do any groups spend more when using a discount than the average non-discount amount?

In [30]:
%%sql
WITH global_baseline AS (
    SELECT age_group,
           ROUND(AVG(purchase_amount), 2) AS baseline_avg_spend
    FROM customer
    WHERE discount_applied = 'No'
    GROUP BY age_group
)
SELECT c.age_group,
       COUNT(c.customer_id) AS total_customers,
       ROUND(AVG(CASE WHEN c.discount_applied = 'Yes' THEN c.purchase_amount END), 2) AS avg_spend_with_discount,
       b.baseline_avg_spend AS baseline_avg_spend_no_discount,
       ROUND(
         AVG(CASE WHEN c.discount_applied = 'Yes' THEN c.purchase_amount END) - b.baseline_avg_spend, 2
       ) AS difference_vs_baseline
FROM customer c
JOIN global_baseline b ON c.age_group = b.age_group
GROUP BY c.age_group, b.baseline_avg_spend
ORDER BY difference_vs_baseline DESC;

age_group,total_customers,avg_spend_with_discount,baseline_avg_spend_no_discount,difference_vs_baseline
Young Adult (18-24),384,61.73,59.17,2.56
Adult (25-34),637,61.08,59.18,1.90
Middle-aged (35-49),1504,61.01,59.32,1.69
Senior (50+),1375,60.80,59.66,1.14


### Is that difference statistically significant?


In [31]:
result = %sql SELECT age_group, discount_applied, purchase_amount FROM customer;
sig_df = result.DataFrame()

from scipy import stats

print("Discount vs. non-discount spend significance check, by age group")
print("-" * 78)
for ag, sub in sig_df.groupby('age_group'):
    disc = sub.loc[sub['discount_applied'] == 'Yes', 'purchase_amount']
    base = sub.loc[sub['discount_applied'] == 'No', 'purchase_amount']
    t, p = stats.ttest_ind(disc, base, equal_var=False)
    sig = "significant (p<0.05)" if p < 0.05 else "not significant"
    print(f"{ag:22s} discount avg={disc.mean():6.2f}  baseline avg={base.mean():6.2f}  p={p:.3f}  -> {sig}")

Discount vs. non-discount spend significance check, by age group
------------------------------------------------------------------------------
Adult (25-34)          discount avg= 61.08  baseline avg= 59.18  p=0.126  -> not significant
Middle-aged (35-49)    discount avg= 61.01  baseline avg= 59.32  p=0.050  -> not significant
Senior (50+)           discount avg= 60.80  baseline avg= 59.66  p=0.206  -> not significant
Young Adult (18-24)    discount avg= 61.73  baseline avg= 59.17  p=0.114  -> not significant


## What are the top 2 most purchased items within each age group?

In [32]:
%%sql
WITH item_counts AS (
    SELECT age_group,
           item_purchased,
           COUNT(*) AS total_orders,
           ROW_NUMBER() OVER (PARTITION BY age_group ORDER BY COUNT(*) DESC) AS item_rank
    FROM customer
    GROUP BY age_group, item_purchased
)
SELECT item_rank,
       age_group,
       item_purchased,
       total_orders
FROM item_counts
WHERE item_rank <= 2
ORDER BY age_group ASC, item_rank ASC;

item_rank,age_group,item_purchased,total_orders
1,Adult (25-34),Jeans,37
2,Adult (25-34),Sunglasses,37
1,Middle-aged (35-49),Socks,74
2,Middle-aged (35-49),Sweater,72
1,Senior (50+),Sunglasses,74
2,Senior (50+),Jacket,67
1,Young Adult (18-24),Shirt,24
2,Young Adult (18-24),Jewelry,22


## What are the top 5 most purchased products within each category?

In [33]:
%%sql
WITH item_counts AS (
    SELECT category,
           item_purchased,
           COUNT(customer_id) AS total_orders,
           ROW_NUMBER() OVER (PARTITION BY category ORDER BY COUNT(customer_id) DESC) AS item_rank
    FROM customer
    GROUP BY category, item_purchased
)
SELECT item_rank, category, item_purchased, total_orders
FROM item_counts
WHERE item_rank <= 5;

item_rank,category,item_purchased,total_orders
1,Accessories,Sunglasses,183
2,Accessories,Hat,178
3,Accessories,Scarf,161
4,Accessories,Handbag,156
5,Accessories,Jewelry,153
1,Clothing,Socks,175
2,Clothing,Pants,169
3,Clothing,Jeans,167
4,Clothing,Sweater,166
5,Clothing,Shorts,158


## What items have the top by Product Rating?

In [34]:
%%sql
SELECT item_purchased,
       ROUND(AVG(review_rating::numeric),2) as "Average Product Rating"
FROM customer
GROUP BY item_purchased
ORDER BY AVG(review_rating) DESC
limit 5

item_purchased,Average Product Rating
T-shirt,3.83
Belt,3.81
Hat,3.80
Socks,3.78
Jewelry,3.78


## Which 5 products have the highest percentage of purchases with discounts applied?

In [35]:
%%sql
SELECT item_purchased,
       ROUND(100.0 * SUM(CASE WHEN discount_applied = 'Yes' THEN 1 ELSE 0 END)/COUNT(*),2) AS discount_rate
FROM customer
GROUP BY item_purchased
ORDER BY discount_rate DESC
LIMIT 5;

item_purchased,discount_rate
Sneakers,52.63
Pants,46.75
Blouse,45.07
Gloves,45.00
Hoodie,44.85


## What Shipping Types have the highest average purchase amount?

In [36]:
%%sql
SELECT shipping_type,
ROUND(AVG(purchase_amount),2) AS "Average Purchase Amount"
FROM customer
GROUP BY shipping_type;

shipping_type,Average Purchase Amount
Express,59.67
Next Day Air,59.54
Store Pickup,59.76
Free Shipping,60.57
2-Day Shipping,60.67
Standard,60.09


### How does revenue vary by season?

In [37]:
%%sql
SELECT season,
       COUNT(*) AS orders,
       ROUND(AVG(purchase_amount),2) AS avg_order_value,
       SUM(purchase_amount) AS total_revenue
FROM customer
GROUP BY season
ORDER BY total_revenue DESC;

season,orders,avg_order_value,total_revenue
Spring,995,60.42,60116
Summer,991,59.45,58918
Winter,969,60.31,58437
Fall,945,60.03,56730


### How does revenue vary by payment method?

In [38]:
%%sql
SELECT payment_method,
       COUNT(*) AS customers,
       ROUND(AVG(purchase_amount),2) AS avg_spend,
       SUM(purchase_amount) AS total_revenue
FROM customer
GROUP BY payment_method
ORDER BY total_revenue DESC;

payment_method,customers,avg_spend,total_revenue
Venmo,648,62.34,40399
Debit Card,674,59.12,39848
Cash,666,59.54,39655
Bank Transfer,649,59.99,38935
Credit Card,637,59.49,37897
PayPal,626,59.85,37467


### Do higher review ratings correlate with customer retention?


In [39]:
%%sql
SELECT customer_segment,
       ROUND(AVG(review_rating::numeric),2) AS avg_review_rating
FROM customer_segments
GROUP BY customer_segment
ORDER BY avg_review_rating DESC;

customer_segment,avg_review_rating
Loyal,3.70
Returning,3.69
New,3.60


### Do discounts or subscriptions correlate with customer retention?

In [40]:
%%sql
SELECT subscription_status,
       discount_applied,
       ROUND(AVG(previous_purchases),1) AS avg_previous_purchases
FROM customer
GROUP BY subscription_status, discount_applied
ORDER BY avg_previous_purchases DESC;

subscription_status,discount_applied,avg_previous_purchases
Yes,No,15.1
No,Yes,15.0
Yes,Yes,14.6
No,No,14.4


### What are the top 10 locations that generate the most revenue?

In [41]:
%%sql
SELECT location,
       COUNT(*) AS customers,
       SUM(purchase_amount) AS total_revenue
FROM customer
GROUP BY location
ORDER BY total_revenue DESC
LIMIT 10;

location,customers,total_revenue
South Carolina,95,5633
Colorado,89,5546
Hawaii,88,5493
New Hampshire,90,5478
New Mexico,87,5471
Kansas,90,5435
Indiana,94,5413
Vermont,92,5409
Utah,88,5363
North Dakota,85,5334


### Channel performance: in-store pickup vs. shipped

In [42]:
%%sql
SELECT CASE WHEN shipping_type = 'Store Pickup' THEN 'In-Store Pickup' ELSE 'Shipped' END AS channel_proxy,
       COUNT(*) AS orders,
       ROUND(AVG(purchase_amount),2) AS avg_spend
FROM customer
GROUP BY channel_proxy;

channel_proxy,orders,avg_spend
Shipped,3250,60.11
In-Store Pickup,650,59.76


## Export Results

In [43]:
load_dotenv()

# Retrieve configuration settings
db_host = os.getenv("DB_HOST")
db_user = os.getenv("DB_USER")
db_password = os.getenv("DB_PASSWORD")
db_name = os.getenv("DB_NAME")
table_name = 'customer'

# Dynamically name the output file using the table name
output_csv = f"{table_name}_export.csv"

# Construct the secure database connection engine
connection_string = f"postgresql://{db_user}:{db_password}@{db_host}:5432/{db_name}"
engine = create_engine(connection_string)

try:
    print(f"Reading data from database table '{table_name}'...")
    df_export = pd.read_sql("SELECT * FROM customer_segments", con=engine)

    df_export.to_csv(output_csv, index=False)
    print(f"Successfully saved data to: {output_csv}")

except Exception as e:
    print(f"An error occurred during the data export: {e}")

Reading data from database table 'customer'...
Successfully saved data to: customer_export.csv


## 📊 Key Findings & Recommendations

**1. Demographic & pattern tracking**
- The customer base skews heavily toward **Loyal** (11+ previous purchases, ~62% of customers) and **Returning** (~37%), with almost no customers in the **New** (1 previous purchase) bucket (<1%). That imbalance is itself a finding: the current customer base looks retention-heavy with very little visible new-customer acquisition.
- Revenue is essentially flat across age groups (~$60 avg spend in every bracket) and only mildly higher for male customers ($60.78 vs. $59.15 avg) — age and gender alone are weak predictors of spend in this data; segmentation strategy likely needs to lean on behavioral fields (subscription, discount usage) instead of demographics alone.

**2. Channel performance**
- Using the `shipping_type`-based proxy, in-store pickup and shipped orders show similar average spend (~$59.76 vs. ~$60.11).

**3. Feature impact: discounts, reviews, payment methods**
- **Discounts:** discount purchases run slightly higher than the non-discount baseline in every age group (+$1.15 to +$1.89), but this difference is **not statistically significant**.
- **Subscriptions:** subscribers spend meaningfully more on average ($65.05 vs. $58.28) and use discounts more often (57% vs. 35% of orders).
- **Reviews:** average rating is nearly flat across lifecycle stages (Loyal 3.70, Returning 3.69, New 3.60), review rating doesn't show a strong link to retention.
- **Payment methods:** revenue is broadly even across payment methods (all within ~$37K–$40K), with no method standing out as a meaningful driver.

**4. Actionable recommendations**
- Prioritize **subscription growth** 
- Investigate the **near-absence of new customers** 
